### ISSUE TITLE
New `NotEmptyRange` and `Bound` types

### ISSUE THREAD

**Issue by LVMVRQUXL** [kotools/types#56]
## Description

Add the following types in the `kotools.types.range` package for using explicit ranges of values:
- the `NotEmptyRange` type representing a range containing at least one value
- the `Bound` type representing the start or the end of a range of values.

Here is an example of the structures:

```kotlin
/** Representation of a range of comparable values. */
public sealed interface NotEmptyRange<T : Comparable<T>> {
    /** The start of this range. */
    public val start: Bound<T>

    /** The end of this range. */
    public val end: Bound<T>

    /**
     * Returns `true` if this range contains the given [value], or returns
     * `false` otherwise.
     */
    public infix operator fun contains(value: T): Boolean

    /** Returns the string representation of this range. */
    override fun toString(): String
}

/** Representation of a bound in a [range][NotEmptyRange]. */
public sealed interface Bound<T : Comparable<T>> {
    /** The value of this bound. */
    public val value: T

    /**
     * Creates a [NotEmptyRange] using this bound and the [other] one.
     * If the [other] bound is lower than this bound, the resulting range will
     * [start][NotEmptyRange.start] with the [other] bound.
     */
    public operator fun rangeTo(other: Bound<T>): NotEmptyRange<T> = TODO()

    /** Returns the string representation of this bound's [value]. */
    override fun toString(): String
}

/** Representation of an inclusive bound in a [range][NotEmptyRange]. */
public data class InclusiveBound<T : Comparable<T>>
internal constructor(override val value: T) : Bound<T>

/** Returns this comparable value as an [InclusiveBound]. */
public fun <T : Comparable<T>> T.toInclusiveBound(): InclusiveBound<T> = TODO()

/** Representation of an exclusive bound in a [range][NotEmptyRange]. */
public data class ExclusiveBound<T : Comparable<T>>
internal constructor(override val value: T) : Bound<T>

/** Returns this comparable value as an [ExclusiveBound]. */
public fun <T : Comparable<T>> T.toExclusiveBound(): ExclusiveBound<T> = TODO()
```

## Checklist

- [x] Implement the `Bound` type.
- [x] Implement the `NotEmptyRange` type.
- [x] Test on all platforms.
- [x] [Refactor](#issuecomment-1501177232).
- [x] Document the new `kotools.types.range` package.
- [x] Update the `Unreleased` section in the changelog.

**Comment by MichaelStH** [comment]
TODO TEST cOMMENT !!! WOOOOW

**Comment by LVMVRQUXL** [comment]
While refactoring code for using these new structures, we should deprecate and replace the `min` and the `max` properties of the following types by a new `range` property:
- [x] `StrictlyPositiveInt`.
- [x] `StrictlyNegativeInt`.
- [x] `PositiveInt`.
- [x] `NegativeInt`.
- [x] `NonZeroInt`.

For the `NonZeroInt`, we should choose between the following:
```kotlin
// Choice 1 : two properties
val negativeRange: NotEmptyRange<StrictlyNegativeInt>
val positiveRange: NotEmptyRange<StrictlyPositiveInt>
// Choice 2 : one property
val ranges: Pair<NotEmptyRange<StrictlyNegativeInt>, NotEmptyRange<StrictlyPositiveInt>>
```

> See [the solution chosen](#issuecomment-1502110535) for the `NonZeroInt` type.

**Comment by LVMVRQUXL** [comment]
> For the `NonZeroInt`, we should choose between the following:
> 
> ```kotlin
> // Choice 1 : two properties
> val negativeRange: NotEmptyRange<StrictlyNegativeInt>
> val positiveRange: NotEmptyRange<StrictlyPositiveInt>
> // Choice 2 : one property
> val ranges: Pair<NotEmptyRange<StrictlyNegativeInt>, NotEmptyRange<StrictlyPositiveInt>>
> ```

We will implement the first choice for readability purpose.

**Comment by LVMVRQUXL** [comment]
For now, we use the following expression for building a not empty range from 1 to 42:

```kotlin
val start = 1
val end = 42
val range: NotEmptyRange<Int> = start.toInclusiveBound()..end.toExclusiveBound()
println(range) // [1;42[
```
The `..` implies that we are creating a range of values by itself.
Also, the bounds used for creating the range doesn't throw any exception when constructing them: they just wrap the given value in a Bound type.

So I propose to use the following expression instead:
```kotlin
val range = notEmptyRangeOf<Int> { 1.inclusive to 42.exclusive }
println(range) // [1;42[
```
The idea is to use properties for building the range's bounds instead of using functions.
This makes the code more concise and idiomatic.

Here's the necessary API for doing so:
```kotlin
fun <T : Comparable<T>> notEmptyRangeOf(
    bounds: NotEmptyRangeBuilderScope<T>.() -> Pair<Bound<T>, Bound<T>>
): NotEmptyRange<T> = TODO()

class NotEmptyRangeBuilderScope<T : Comparable<T>> {
    val T.inclusive: InclusiveBound<T> get() = TODO()
    val T.exclusive: ExclusiveBound<T> get() = TODO()
}
```

**Comment by LVMVRQUXL** [comment]
We should also fix the issues reported by this [comment](https://github.com/kotools/types/issues/73#issuecomment-1547433012).

**Comment by LVMVRQUXL** [comment]
Here's a new proposal for a minimal API of the `NotEmptyRange` and the `Bound` types:
```kotlin
sealed interface Bound<out T : Comparable<@UnsafeVariance T>> {
    val value: T
    override fun toString(): String
}

interface InclusiveBound<out T : Comparable<@UnsafeVariance T>> : Bound<T>
interface ExclusiveBound<out T : Comparable<@UnsafeVariance T>> : Bound<T>

fun <T : Comparable<T>> notEmptyRangeOf(
    bounds: NotEmptyRange.BuilderScope<T>.() -> Pair<Bound<T>, Bound<T>>
): NotEmptyRange<T> = TODO()

interface NotEmptyRange<out T : Comparable<@UnsafeVariance T>> {
    val start: Bound<T>
    val end: Bound<T>
    override fun toString(): String

    class BuilderScope<T : Comparable<T>> {
        val T.inclusive: InclusiveBound<T> get() = TODO()
        val T.exclusive: ExclusiveBound<T> get() = TODO()
    }
}
```

This API could be used like this:
```kotlin
val range = notEmptyRangeOf<Int> { 1.inclusive to 42.exclusive }
println(range) // [1;42[
```

**Comment by LVMVRQUXL** [comment]
Like defined in our [strategy](https://github.com/kotools/types/discussions/25?sort=old#discussioncomment-5909618), we should introduce these declarations in **experimental** stage.
Because of this decision, we shouldn't deprecate the `min` and the `max` properties of the `StrictlyPositiveInt`, the `StrictlyNegativeInt`, the `PositiveInt`, the `NegativeInt` and the `NonZeroInt` types.

**Comment by LVMVRQUXL** [comment]
Here's a summary of all tasks to do for resolving this issue:

- [x] Add the `ExperimentalRangeApi` annotation in the `kotools.types.experimental` package.
- [x] Remove the deprecation notice on the `min` and the `max` properties of the `StrictlyPositiveInt`, the `StrictlyNegativeInt`, the `PositiveInt`, the `NegativeInt` and the `NonZeroInt` types.
- [x] Implement the [experimental API](https://github.com/kotools/types/issues/56#issuecomment-1547602384) in the `kotools.types.experimental` package.
- [x] Remove the old implementation shipped in the `4.2.0-alpha` version.
- [x] Move the experimental API to the `kotools.types.range` package.
- [x] Refactor.
- [x] Update the `Unreleased` section in the changelog.
